idx/memory/tag_query.go

```diff
@@ -0,0 +1,411 @@
+package memory
+
+import (
+	"errors"
+	"math"
+	"regexp"
+
+	"github.com/grafana/metrictank/idx"
+	"github.com/raintank/worldping-api/pkg/log"
+)
+
+var (
+	errInvalidQuery = errors.New("invalid query")
+)
+
+const (
+	PARSING_ERROR = iota
+	EQUAL
+	NOT_EQUAL
+	MATCH
+	NOT_MATCH
+)
+
+type kv struct {
+	key   string
+	value string
+}
+
+type expression struct {
+	kv
+	operator int
+}
+
+type TagQuery struct {
+	from     int64
+	equal    []kv
+	match    []kv
+	notEqual []kv
+	notMatch []kv
+}
+
+// parseExpression returns an expression that's been generated from the given
+// string, in case of error the operator will be PARSING_ERROR.
+func parseExpression(expr string) expression {
+	var pos int
+	regex, not := false, false
+
+	// scan up to operator to get key
+	for ; pos < len(expr); pos++ {
+		// ! || =
+		if expr[pos] == 33 || expr[pos] == 61 {
+			// key must not be empty
+			if pos == 0 {
+				return expression{operator: PARSING_ERROR}
+			}
+			break
+		}
+
+		// disallow ; in key
+		if expr[pos] == 59 {
+			return expression{operator: PARSING_ERROR}
+		}
+	}
+
+	key := expr[:pos]
+
+	// if !
+	if expr[pos] == 33 {
```

Comment:
this will generate a panic when pos is out of bounds. (happens when we don't break out of the loop, eg for expr = `foo`)

Reply:
actually, i would make 2 separate clauses in the loop, one for 33, one for 61. you can then move this logic in there, instead of doing the same check again.

Reply:
like that? https://github.com/grafana/metrictank/pull/729/commits/1e2fe069f28087dc915c497a61aca4cb7517168d

Performance is similar:
```
before:
BenchmarkExpressionParsing-8   	20000000	       913 ns/op	     309 B/op	      12 allocs/op

after:
BenchmarkExpressionParsing-8   	20000000	       904 ns/op	     309 B/op	      12 allocs/op
```